File: scripts/run_trigger_eval.py
#!/usr/bin/env python3
"""Run trigger cases through an optional JSON-lines evaluator command."""

from __future__ import annotations

import argparse
import json
import shlex
import subprocess
from datetime import UTC, datetime

from scripts.trigger_eval import ROOT, build_matrix


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", default="evals/results/latest.json")
    parser.add_argument("--limit", type=int)
    parser.add_argument("--sample", type=int)
    parser.add_argument("--command")
    args = parser.parse_args()
    matrix = build_matrix()
    work = [(h, e, c) for h in matrix["hosts"] for e in matrix["environments"] for c in matrix["cases"]]
    if args.limit is not None:
        work = work[: args.limit]
    if args.sample is not None and args.sample < len(work):
        work = [work[int(index * len(work) / args.sample)] for index in range(args.sample)]
    results = [_evaluate(host, env, case, args.command) for host, env, case in work]
    counts = {status: sum(row["status"] == status for row in results) for status in ("passed", "failed", "not_run")}
    document = {
        "schema_version": 1,
        "generated_at": datetime.now(UTC).isoformat(),
        "summary": {"total": len(results), **counts},
        "results": results,
    }
    output = (ROOT / args.output).resolve()
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(document, ensure_ascii=False, indent=2) + "\n")
    print(f"actual trigger evaluation: {counts} ({len(results)} total)")
    return int(counts["failed"] > 0)


def _evaluate(host: str, env: str, case: dict, command: str | None) -> dict:
    base = {"host": host, "environment": env, "skill": case["skill"], "type": case["type"], "case_id": case["id"]}
    if not command:
        return {**base, "status": "not_run", "actual": None, "error": "no evaluator command supplied"}
    process = subprocess.run(
        shlex.split(command),
        input=json.dumps({"host": host, "environment": env, "case": case}),
        text=True,
        capture_output=True,
        check=False,
    )
    try:
        actual = json.loads(process.stdout)["selected_skill"]
    except (json.JSONDecodeError, KeyError) as exc:
        return {**base, "status": "failed", "actual": None, "error": f"{exc}; stderr={process.stderr}"}
    passed = process.returncode == 0 and actual == case["expected"]
    return {**base, "status": "passed" if passed else "failed", "actual": actual, "error": process.stderr or None}


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/run_trigger_eval.rb
#!/usr/bin/env ruby
require "json"
require "open3"
require "optparse"
require "time"

options = {output: "evals/results/latest.json", limit: nil, sample: nil, command: nil}
OptionParser.new do |parser|
  parser.on("--output PATH") { |value| options[:output] = value }
  parser.on("--limit N", Integer) { |value| options[:limit] = value }
  parser.on("--sample N", Integer) { |value| options[:sample] = value }
  parser.on("--command COMMAND") { |value| options[:command] = value }
end.parse!

root = File.expand_path("..", __dir__)
cases_json, status = Open3.capture2("ruby", File.join(root, "scripts/trigger_eval.rb"), "--json")
abort("failed to build trigger cases") unless status.success?
matrix = JSON.parse(cases_json)
work = matrix.fetch("hosts").product(matrix.fetch("environments"), matrix.fetch("cases"))
work = work.first(options[:limit]) if options[:limit]
if options[:sample] && options[:sample] < work.length
  work = options[:sample].times.map { |index| work[(index * work.length / options[:sample].to_f).floor] }
end

results = work.map do |host, environment, kase|
  base = {host: host, environment: environment, skill: kase.fetch("skill"), type: kase.fetch("type"), case_id: kase.fetch("id")}
  unless options[:command]
    next base.merge(status: "not_run", actual: nil, error: "no evaluator command supplied")
  end
  payload = JSON.generate({host: host, environment: environment, case: kase})
  stdout, stderr, process = Open3.capture3(options[:command], stdin_data: payload)
  begin
    actual = JSON.parse(stdout).fetch("selected_skill")
    expected = kase.fetch("expected")
    base.merge(status: process.success? && actual == expected ? "passed" : "failed", actual: actual, error: stderr.empty? ? nil : stderr)
  rescue => error
    base.merge(status: "failed", actual: nil, error: "#{error.message}; stderr=#{stderr}")
  end
end

counts = %w[passed failed not_run].to_h { |key| [key, results.count { |result| result[:status] == key }] }
document = {schema_version: 1, generated_at: Time.now.utc.iso8601, summary: {total: results.length, **counts.transform_keys(&:to_sym)}, results: results}
path = File.expand_path(options[:output], root)
Dir.mkdir(File.dirname(path)) unless Dir.exist?(File.dirname(path))
File.write(path, JSON.pretty_generate(document) + "\n")
puts "actual trigger evaluation: #{counts['passed']} passed, #{counts['failed']} failed, #{counts['not_run']} not_run (#{results.length} total)"
exit 1 if counts["failed"] > 0
